pkg/kademlia/pslice/pslice_test.go

```diff
@@ -0,0 +1,309 @@
+
+// TestAddRemove checks that the Add, Remove and Exists methods work as expected.
+func TestAddRemove(t *testing.T) {
+	var (
+		ps    = pslice.New(4)
+		base  = test.RandomAddress()
+		peers = make([]swarm.Address, 8)
+	)
+
+	// 2 peers per bin
+	// indexes {0,1} {2,3} {4,5} {6,7}
+	for i := 0; i < 8; i += 2 {
+		a := test.RandomAddressAt(base, i)
+		peers[i] = a
+
+		b := test.RandomAddressAt(base, i)
+		peers[i+1] = b
+	}
+	chkNotExists(t, ps, peers...)
+
+	// add one
+	ps.Add(peers[0], 0)
+	chkLen(t, ps, 1)
+	chkExists(t, ps, peers[:1]...)
+	chkNotExists(t, ps, peers[1:]...)
+
+	// check duplicates
+	ps.Add(peers[0], 0)
+	chkLen(t, ps, 1)
+	chkBins(t, ps, []uint{0, 1, 1, 1})
+	chkExists(t, ps, peers[:1]...)
+	chkNotExists(t, ps, peers[1:]...)
+
+	// check empty
+	ps.Remove(peers[0], 0)
+	chkLen(t, ps, 0)
+	chkBins(t, ps, []uint{0, 0, 0, 0})
+	chkNotExists(t, ps, peers...)
+
+	// add two in bin 0
+	ps.Add(peers[0], 0)
+	ps.Add(peers[1], 0)
+	chkLen(t, ps, 2)
+	chkBins(t, ps, []uint{0, 2, 2, 2})
+	chkExists(t, ps, peers[:2]...)
+	chkNotExists(t, ps, peers[2:]...)
+
+	ps.Add(peers[2], 1)
+	ps.Add(peers[3], 1)
+	chkLen(t, ps, 4)
+	chkBins(t, ps, []uint{0, 2, 4, 4})
+	chkExists(t, ps, peers[:4]...)
+	chkNotExists(t, ps, peers[4:]...)
+
+	ps.Remove(peers[1], 0)
+	chkLen(t, ps, 3)
+	chkBins(t, ps, []uint{0, 1, 3, 3})
+	chkExists(t, ps, peers[0], peers[2], peers[3])
+	chkNotExists(t, ps, append([]swarm.Address{peers[1]}, peers[4:]...)...)
+
+	// this should not move the last cursor
+	ps.Add(peers[7], 3)
+	chkLen(t, ps, 4)
+	chkBins(t, ps, []uint{0, 1, 3, 3})
+	chkExists(t, ps, peers[0], peers[2], peers[3], peers[7])
+	chkNotExists(t, ps, append([]swarm.Address{peers[1]}, peers[4:7]...)...)
+
+	ps.Add(peers[5], 2)
+	chkLen(t, ps, 5)
+	chkBins(t, ps, []uint{0, 1, 3, 4})
+	chkExists(t, ps, peers[0], peers[2], peers[3], peers[5], peers[7])
+	chkNotExists(t, ps, []swarm.Address{peers[1], peers[4], peers[6]}...)
+
+	ps.Remove(peers[2], 1)
+	chkLen(t, ps, 4)
+	chkBins(t, ps, []uint{0, 1, 2, 3})
+	chkExists(t, ps, peers[0], peers[3], peers[5], peers[7])
+	chkNotExists(t, ps, []swarm.Address{peers[1], peers[2], peers[4], peers[6]}...)
+
+	p := uint8(0)
+	for i := 0; i < 8; i += 2 {
+		ps.Remove(peers[i], p)
+		ps.Remove(peers[i+1], p)
+		p++
+	}
+
+	// check empty again
+	chkLen(t, ps, 0)
+	chkBins(t, ps, []uint{0, 0, 0, 0})
+	chkNotExists(t, ps, peers...)
+}
+
+// TestIteratorError checks that error propagation works correctly in the iterators.
+func TestIteratorError(t *testing.T) {
+	var (
+		ps   = pslice.New(4)
+		base = test.RandomAddress()
+		a    = test.RandomAddressAt(base, 0)
+		e    = errors.New("err1")
+	)
+
+	ps.Add(a, 0)
+
+	f := func(p swarm.Address, _ uint8) (stop, jumpToNext bool, err error) {
+		return false, false, e
+	}
+
+	err := ps.EachBin(f)
+	if !errors.Is(err, e) {
+		t.Fatal("didnt get expected error")
+	}
+}
+
+// TestIterators tests that the EachBin and EachBinRev iterators work as expected.
+func TestIterators(t *testing.T) {
```

Comment:
It looks to me that stop and jumpToNext functionalities are not tested here.

Reply:
added tests for that :+1: there was a bug there. nice catch :+1: :clap: